### PR TITLE
Form Validation and responsiveness for Doser

### LINKED ISSUE #579: Update Doser UI
Add UI validation to Doser tab and update to responsive design.

### PR DESCRIPTION
**Pull request by Ranthalion**
- Add validation and responsiveness to Doser UI
- Update Confirm's cancel button style
- Move doser calibration to modal
- Tweak Collapsible to stay expanded during edit mode
- Updated Doser API functionality to maintain cron schedule during Create and Update.
Fixes #579 

Preview available at http://reef-pi.herokuapp.com

image
image
image

### CODE REVIEW

File: jsx/doser/calibrate.jsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import React from 'react'
+import * as Yup from 'yup'
+import { ErrorFor, ShowError } from '../utils/validation_helper'
+import classNames from 'classnames'
+import { withFormik, Field } from 'formik'
+
+export const Calibrate = ({
```
Comment (zekth) [inline jsx/doser/calibrate.jsx]: Why not using the class pattern for component like everywhere in the UI? Makes it more consistent.

File: jsx/doser/edit_doser.jsx [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  readOnly
+}) => {
+  const handleSubmit = event => {
+    event.preventDefault()
+    clearAlert()
+    if (dirty === false || isValid === true) {
+      submitForm()
+    } else {
+      submitForm() // Calling submit form in order to show validation errors
+      showAlert(
```
Comment (zekth) [inline jsx/doser/edit_doser.jsx]: Use the dispatch nofication because this call is deprecated

Comment (ranjib) [reply jsx/doser/edit_doser.jsx]: Do you mean addLog ? The main panel has notification component already

File: jsx/doser/calibrate.jsx (hunk 1 above)
Comment (ranjib) [reply jsx/doser/calibrate.jsx]: It’s ok to have mixed style if we don’t have any state based logic, the react style guide recommends this as well. I remember asking the same to @Ranthalion

Comment (Ranthalion) [reply jsx/doser/calibrate.jsx]: In react, it is preferred to use stateless function if the component does not need to track state.  Some linters even include this as a rule, but the one we are using doesn't seem to include it.

File: jsx/doser/edit_doser.jsx (hunk 2 above)
Comment (Ranthalion) [reply jsx/doser/edit_doser.jsx]: The dispatch call is pretty long.  What about keeping showAlert as a convenience method instead?  This also removes the need to rewrite a bunch of the other code that is already using showAlert.

Comment (Ranthalion) [reply jsx/doser/edit_doser.jsx]: This also places an entry in the notification log.  In this case, the alert is specifically to communicate with the user; it should not be persisted in the log.  It is not a system error, it is a validation failure.
Maybe the right answer is to refactor utils/alert to provide functionality that makes it easy to show info/warning/error and log to notification as appropriate.  I'll think on this and add comments to the issue posted earlier.

Comment (zekth) [reply jsx/doser/edit_doser.jsx]: https://github.com/reef-pi/reef-pi/blob/master/jsx/utils/ajax.js#L27 like here but there is a method to log only and no display. May i add an entry to show notification and don't log it?

File: jsx/doser/calibrate.jsx (hunk 1 above)
Comment (zekth) [reply jsx/doser/calibrate.jsx]: So do you suggest me to migrate in this pattern for Notification Alerts?

Comment (Ranthalion) [reply jsx/doser/calibrate.jsx]: I don't think it's a big deal.  I quickly looked for any benchmarks that show performance differences, but didn't really find anything compelling.  
Functional components vs Stateful components was a topic when I went to a Reactjs meetup.

File: jsx/doser/edit_doser.jsx (hunk 2 above)
Comment (Ranthalion) [reply jsx/doser/edit_doser.jsx]: See Issue #590 for new comments and additional conversation.

**Pull request merged**